Show errant values in current-domain out-of-bounds error message

johnkerl opened the pull request:
<long description>

[[sc-61579]](https://app.shortcut.com/tiledb-inc/story/61579/print-actual-coordinate-values-for-out-of-bounds-errors)

Before:

```
tiledbsoma._exception.SOMAError: [ManagedQuery] [unnamed] Query FAILED: Query: A range was set outside the current domain.
```

After:

```
tiledbsoma._exception.SOMAError: [ManagedQuery] [unnamed] Query FAILED: Query: A range [99999, 99999] on dimension 'soma_joinid' was set outside of the current domain [[0, 2637]].
```

---
TYPE: IMPROVEMENT
DESC: Show errant values in current-domain out-of-bounds error message

ihnorton commented on tiledb/sm/query/query.cc:
```diff
@@ -859,8 +860,13 @@ Status Query::process() {
+                  "domain {}.",
+                  range_str(
+                      range, array_schema_->domain().dimension_ptr(d)->type()),
+                  array_schema_->domain().dimension_ptr(d)->name(),
```
Shouldn't this be `current_domain()`?

johnkerl replied:
We need the domain to get types. Those are wired through `array_schema_` and `current_domain` both. I can get them either way.

johnkerl replied:
@ihnorton correction: the domain is wired through the `array_schema_` and the private `ndrectangle_` of `CurrentDomain`.

It is an invariant that they be the same domain in both places.

If you want to pursue the latter then we need to connect a domain accessor in `CurrentDomain` which passes through the copy of the domain in its `ndrectangle_`. Personally I don't see the benefit to doing the latter.

johnkerl commented:
Even though we moved to C++ 20 a long time ago, it appears that in CI there are lots of builds using C++ 17. And I've introduced `std::format` on this PR.

@ihnorton @teo-tsirpanis your thoughts please:

* I can use `fmt::format`
* I can `std::stringstream ss`, `ss << "this"`, `ss << "that"`, etc. & then `throw QueryException(ss.str())`

teo-tsirpanis commented on tiledb/sm/query/query.cc:
```diff
@@ -59,6 +59,7 @@
 #include "tiledb/sm/tile/writer_tile_tuple.h"
 
 #include <cassert>
+#include <format>
```
We unfortunately can't use `<format>` because it breaks CI in macOS and Linux.

teo-tsirpanis commented on tiledb/sm/array_schema/current_domain.cc:
```diff
@@ -408,6 +408,16 @@ void CurrentDomain::expand_to_tiles(
+  if (type_ == CurrentDomainType::NDRECTANGLE) {
+    return ndrectangle_->as_string();
+  } else {
+    // As of 2025-01-09 there is no other such type. When/if we do make such a
+    // type, we'd need to configure it to return a description of itself.
+    return "CurrentDomain of non-NDRectangle type";
```
I would assert or throw here.

teo-tsirpanis commented:
> it appears that in CI there are lots of builds using C++ 17

That's definitely not the case, we use C++20 language features like concepts without any problem. It looks like the compiler versions we are using do not have the library implemented.

Better use fmt instead.

johnkerl commented:
Thank you @teo-tsirpanis !!
